File: GeeksforGeeks_Solutions/C++/October_04.cpp
#include <iostream>
using namespace std;

class Solution{
public:
    // Function to reverse a circular linked list
    Node* reverse(Node* head) {
        Node *i=head;
        Node *nxt;
        Node *pre=NULL;
        do{
            nxt=i->next;
            i->next=pre;
            pre=i;
            i=nxt;
        }while(i!=head);
        i->next=pre;
        head=pre;
        return head;
    }

    // Function to delete a node from the circular linked list
    Node* deleteNode(Node* head, int key) {
        if(!head)return head;
        Node *i=head;
        while(i->data!=key){
            i=i->next;
            if(i==head)break;
        }
        if(i->data==key){
            i->data=i->next->data;
            i->next=i->next->next;
        }
        return head;
    }
};

File: GeeksforGeeks_Solutions/python/October_04.py
class Solution:

    def reverse(self, head):
        if not head:
            return None
        
        prev = None
        current = head
        tail = head          
        
        while True:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
            
            if current == head:
                break
        
        tail.next = prev
        return prev  



    def deleteNode(self, head, key):
        if not head:
            return None
        
        current = head
        prev = None
        
        if current.data == key and current.next == head:
            return None  
        
        while True:
            if current.data == key:
                if prev:
                    prev.next = current.next
                else:
                    while current.next != head:
                        current = current.next
                    current.next = head.next  
                    head = head.next  
                break
            
            prev = current
            current = current.next
            
            if current == head:
                break
        
        return head